### PR TITLE
vApp template capture support

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds functionality to create vApp template from existing vApp or standalone VMs (which also have vApps, they're just hidden). The functionality is called "capture vApp template" and the functionality is introduced using `Catalog.CaptureVappTemplate`, `Catalog.CaptureVappTemplateAsync` and type `types.CaptureVAppParams`

Additionally there are some covenience changes in the code:
* `VApp.GetParentVDC` becomes public (previously it was private method)
* New method `Org.GetVAppByHref` to retrieve a vApp by given HREF
* New methods `VAppTemplate.GetCatalogItemHref` and `VAppTemplate.GetCatalogItemId` that can return
  related catalog item ID and HREF of a vApp template

### CODE REVIEW

File: govcd/common_test.go [hunk 1]
```diff
@@ -856,3 +856,80 @@ func (vcd *TestVCD) checkSkipWhenApiToken(check *C) {
 		check.Skip("This test can't run on API token")
 	}
 }
+
+func createNsxtVAppAndVm(vcd *TestVCD, check *C) (*VApp, *VM) {
```
Comment (Didainius) [inline govcd/common_test.go:860]: Nothing has changed. Moved this chunk of code to avoid issues with test tags.

**Review by lvirbalas — APPROVED** [review]
Thanks!

File: govcd/catalog_test.go [hunk 2]
```diff
@@ -1490,3 +1490,40 @@ func (vcd *TestVCD) Test_CatalogCreateCompleteness(check *C) {
...
+		Source: &types.Reference{
+			HREF: vapp.VApp.HREF,
+		},
+		CustomizationSection: types.CaptureVAppParamsCustomizationSection{
+			Info:                   "CustomizeOnInstantiate Settings",
+			CustomizeOnInstantiate: true,
+		},
+	}
+
+	if vcd.client.Client.APIVCDMaxVersionIs(">= 37.2") {
```
Comment (adambarreiro) [inline govcd/catalog_test.go:1517]: I was having the idea of "hiding" this logic inside `CaptureVappTemplate` so it is done "automatically", but not sure if that would be "too much" (I mean, if <37.2 returns a meaningful error maybe it's not so convenient to hide it)

**Review by adambarreiro — COMMENTED** [review]
First scan! Just some thoughts

File: govcd/catalog_test.go (hunk 2 above)
Comment (Didainius) [reply govcd/catalog_test.go:1517]: I am up for both options. The error is not very beautiful, but it tells that it is an unrecognised field. I could migrate it.

File: .changes/v2.23.0/652-improvements.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+* New public method `VApp.GetParentVDC` to retrieve parent VDC of vApp (previously it was private)
```
[the remark this answers is not in the document]
Comment (Didainius) [reply .changes/v2.23.0/652-improvements.md:1]: Changed

File: govcd/catalog.go [hunk 4]
```diff
@@ -360,6 +360,37 @@ func (cat *Catalog) UploadOvfByLink(ovfUrl, itemName, description string) (Task,
...
+	// After the task is finished, owner field contains the resulting vApp template
+	return cat.GetVappTemplateByHref(task.Task.Owner.HREF)
+}
+
+// CaptureVappTemplateAsync triggers vApp template capturing task and returns it
+func (cat *Catalog) CaptureVappTemplateAsync(captureParams *types.CaptureVAppParams) (Task, error) {
+	util.Logger.Printf("[TRACE] Capturing vApp template to catalog %s", cat.Catalog.Name)
+	captureTemplateHref := cat.client.VCDHREF
+	captureTemplateHref.Path += fmt.Sprintf("/catalog/%s/action/captureVApp", extractUuid(cat.Catalog.ID))
+
```
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/catalog.go:384]: Added

File: govcd/catalog_test.go (hunk 2 above)
Comment (Didainius) [reply govcd/catalog_test.go:1517]: Changed.

File: .changes/v2.23.0/652-improvements.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (adambarreiro) [reply .changes/v2.23.0/652-improvements.md]: Missing the other entries of changelog as well.

**Review by adambarreiro — APPROVED** [review]
LGTM

File: .changes/v2.23.0/652-improvements.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+* Added public method `VApp.GetParentVDC` to retrieve parent VDC of vApp (previously it was private)
+  [GH-652]
+* New methods `Catalog.CaptureVappTemplate`, `Catalog.CaptureVappTemplateAsync` and type
```
Comment (dataclouder) [inline .changes/v2.23.0/652-improvements.md:3]:
```suggestion
* Added methods `Catalog.CaptureVappTemplate`, `Catalog.CaptureVappTemplateAsync` and type
```

File: .changes/v2.23.0/652-improvements.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  `types.CaptureVAppParams` that add support for creating catalog template from existing vApp
+  [GH-652]
+* New method `Org.GetVAppByHref` to retrieve a vApp by given HREF [GH-652]
+* New methods `VAppTemplate.GetCatalogItemHref` and `VAppTemplate.GetCatalogItemId` that can return
```
Comment (dataclouder) [inline .changes/v2.23.0/652-improvements.md:7]:
```suggestion
* Added method `Org.GetVAppByHref` to retrieve a vApp by given HREF [GH-652]
* Added methods `VAppTemplate.GetCatalogItemHref` and `VAppTemplate.GetCatalogItemId` that can return
```

File: govcd/catalog.go [hunk 7]
```diff
@@ -360,6 +360,49 @@ func (cat *Catalog) UploadOvfByLink(ovfUrl, itemName, description string) (Task,
...
+		return nil, err
+	}
+
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, err
+	}
+
+	// After the task is finished, owner field contains the resulting vApp template
+	return cat.GetVappTemplateByHref(task.Task.Owner.HREF)
```
Comment (dataclouder) [inline govcd/catalog.go:376]: Should we test for `task.Task.Owner` to exist before returning?
We risk a panic if the task had an accident and lost the owner

Comment (Didainius) [reply govcd/catalog.go:380]: Patched and ran test

File: .changes/v2.23.0/652-improvements.md (hunk 6 above)
Comment (Didainius) [reply .changes/v2.23.0/652-improvements.md]: Changed.

File: .changes/v2.23.0/652-improvements.md (hunk 5 above)
Comment (Didainius) [reply .changes/v2.23.0/652-improvements.md]: Changed.

**Review by dataclouder — APPROVED** [review]

**Pull request merged**